File: app/frontend/src/theme.ts
export const THEME_KEYS = [
  'constructos-light',
  'constructos-night',
  'symphony-light',
  'symphony-night',
] as const

export type ThemeKey = (typeof THEME_KEYS)[number]
export type ThemeMode = 'light' | 'dark'
export type ThemeBrand = 'constructos' | 'symphony'

export const THEME_LABELS: Record<ThemeKey, string> = {
  'constructos-light': 'ConstructOS light',
  'constructos-night': 'ConstructOS night',
  'symphony-light': 'Symphony light',
  'symphony-night': 'Symphony night',
}

const LEGACY_THEME_ALIASES: Record<string, ThemeKey> = {
  light: 'constructos-light',
  dark: 'constructos-night',
}

const DEFAULT_THEME: ThemeKey = 'symphony-light'

export function normalizeTheme(value: unknown): ThemeKey {
  const normalized = String(value || '').trim().toLowerCase()
  if ((THEME_KEYS as readonly string[]).includes(normalized)) return normalized as ThemeKey
  if (Object.prototype.hasOwnProperty.call(LEGACY_THEME_ALIASES, normalized)) {
    const aliased = LEGACY_THEME_ALIASES[normalized as keyof typeof LEGACY_THEME_ALIASES]
    if (aliased) return aliased
  }
  return DEFAULT_THEME
}

export function getThemeMode(theme: ThemeKey): ThemeMode {
  return theme.endsWith('-night') ? 'dark' : 'light'
}

export function getThemeBrand(theme: ThemeKey): ThemeBrand {
  return theme.startsWith('symphony-') ? 'symphony' : 'constructos'
}

export function toggleTheme(theme: ThemeKey): ThemeKey {
  const brand = getThemeBrand(theme)
  const mode = getThemeMode(theme)
  if (brand === 'symphony') return mode === 'dark' ? 'symphony-light' : 'symphony-night'
  return mode === 'dark' ? 'constructos-light' : 'constructos-night'
}

File: app/shared/theme.py
from __future__ import annotations

from typing import Final

THEME_CONSTRUCTOS_LIGHT: Final[str] = "constructos-light"
THEME_CONSTRUCTOS_NIGHT: Final[str] = "constructos-night"
THEME_SYMPHONY_LIGHT: Final[str] = "symphony-light"
THEME_SYMPHONY_NIGHT: Final[str] = "symphony-night"

DEFAULT_THEME: Final[str] = THEME_SYMPHONY_LIGHT

VALID_THEMES: Final[frozenset[str]] = frozenset(
    {
        THEME_CONSTRUCTOS_LIGHT,
        THEME_CONSTRUCTOS_NIGHT,
        THEME_SYMPHONY_LIGHT,
        THEME_SYMPHONY_NIGHT,
    }
)

LEGACY_THEME_ALIASES: Final[dict[str, str]] = {
    "light": THEME_CONSTRUCTOS_LIGHT,
    "dark": THEME_CONSTRUCTOS_NIGHT,
}


def normalize_theme(value: object, *, default: str = DEFAULT_THEME) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in VALID_THEMES:
        return normalized
    if normalized in LEGACY_THEME_ALIASES:
        return LEGACY_THEME_ALIASES[normalized]
    return default


def theme_mode(theme: object) -> str:
    normalized = normalize_theme(theme)
    return "dark" if normalized.endswith("-night") else "light"


def toggle_theme(theme: object) -> str:
    normalized = normalize_theme(theme)
    if normalized.startswith("symphony-"):
        return THEME_SYMPHONY_LIGHT if normalized.endswith("-night") else THEME_SYMPHONY_NIGHT
    return THEME_CONSTRUCTOS_LIGHT if normalized.endswith("-night") else THEME_CONSTRUCTOS_NIGHT
